amethyst_assets/src/stream.rs

```diff
@@ -0,0 +1,319 @@
+            }
+            COMPLETE => {
+                return Ok(Async::Ready(None));
+            }
+            POISONED => panic!("inner stream panicked during poll"),
+            _ => unreachable!(),
+        }
+
+        loop {
+            struct Reset<'a>(&'a AtomicUsize);
+
+            impl<'a> Drop for Reset<'a> {
+                fn drop(&mut self) {
+                    use std::thread;
+
+                    if thread::panicking() {
+                        self.0.store(POISONED, SeqCst);
+                    }
+                }
+            }
+
+            let _reset = Reset(&self.inner.notifier.state);
+
+            // Poll the stream
+            let res = unsafe {
+                (*self.inner.stream.get())
+                    .as_mut()
+                    .unwrap()
+                    .poll_stream_notify(&self.inner.notifier, 0)
+            };
+            match res {
+                Ok(Async::NotReady) => {
+                    // Not ready, try to release the handle
+                    match self.inner
+                        .notifier
+                        .state
+                        .compare_and_swap(POLLING, IDLE, SeqCst)
+                    {
+                        POLLING => {
+                            // Success
+                            return Ok(Async::NotReady);
+                        }
+                        REPOLL => {
+                            // Gotta poll again!
+                            let prev = self.inner.notifier.state.swap(POLLING, SeqCst);
+                            assert_eq!(prev, REPOLL);
+                        }
+                        _ => unreachable!(),
+                    }
+                }
+                Ok(Async::Ready(i)) => {
+                    let complete = i.is_none();
+                    (*self.inner.last.write()) = Ok(Async::Ready(i));
+                    self.inner.counter.fetch_add(1, Ordering::Release);
+                    if complete {
+                        self.complete();
+                    } else {
+                        self.idle();
+                    }
+                    break;
+                }
+                Err(e) => {
+                    (*self.inner.last.write()) = Err(SharedError { error: Arc::new(e) });
+                    self.inner.counter.fetch_add(1, Ordering::Release);
+
+                    self.complete();
+                    break;
+                }
+            }
+        }
+
+        self.peek()
+    }
+}
+
+impl<S> Clone for UpdatesStream<S>
+where
+    S: Stream,
+{
+    fn clone(&self) -> Self {
+        let next_clone_id = self.inner.next_clone_id.fetch_add(1, SeqCst);
+
+        UpdatesStream {
+            inner: self.inner.clone(),
+            waiter: next_clone_id,
+            counter: 0,
+        }
+    }
+}
+
+impl<S> Drop for UpdatesStream<S>
+where
+    S: Stream,
+{
+    fn drop(&mut self) {
+        let mut waiters = self.inner.notifier.waiters.lock().unwrap();
```

Comment:
We need tests for this sort of stuff!

Reply:
I've copied it from `futures` crate. Should be fine :smile:
`futures::future::Shared` is doing the same and we use them.
I have an idea for making all this much simplier and maybe faster but I need linear types to enforce required contract on type level.